cmaes.lua

```diff
@@ -0,0 +1,273 @@
+require 'torch'
+require 'math'
+
+local BestSolution = {} 
+--[[ An implementation of `CMAES` (Covariance Matrix Adaptation Evolution Strategy), 
+ported from https://www.lri.fr/~hansen/barecmaes2.html.
+ 
+Parameters
+----------
+ARGS:
+
+-    `opfunc` : a function that takes a single input (X), the point of 
+          evaluation, and returns f(X) and df/dX. Note that df/dX is not used 
+-    `x` : the initial point
+-    `state.sigma`
+        float, initial step-size (standard deviation in each
+        coordinate)
+-    `state.maxEval`
+        int, maximal number of function evaluations
+-    `state.ftarget`
+        float, target function value
+-    `state.popsize`
+    	 population size. If this is left empty, 
+    	  4 + int(3 * log(|x|)) will be used
+-    `state.ftarget` 
+        stop if fitness < ftarget
+-    `state.verd_disp`
+        int, display on console every verb_disp iteration, 0 for never
+-    `state.args`
+        arguments to `opfunc`
+
+RETURN:
+- `x*` : the new `x` vector, at the optimal point
+- `f`  : a table of all function values: 
+     `f[1]` is the value of the function before any optimization and
+     `f[#f]` is the final fully optimized value, at `x*`
+--]]
+function optim.cmaes(opfunc, x, config, state)
+   -- process input parameters
+   local config = config or {}
+   local state = state or config
+   local xmean = torch.Tensor(x:clone():double():storage()) -- distribution mean, a flattened copy
```

Comment:
The problem with using x:clone():view(-1) is, that the Tensor might be of a different type than the default tensor. The internal tensors of cmaes are DoubleTensors by default, which will lead to type mismatch  errors in several operations  (already floatTensors as x don't work). Typing all internal variables as the type of x is also not possible because some Tensors (e.g. cudatensor) don't have all operations implemented. 
Is there a better way to do this ?